File: ARRAY/2D_Array/TwoD_Array.java
import java.util.Scanner;

public class TwoD_Array {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Asking the user for rows and columns
        System.out.print("Enter the number of rows: ");
        int rows = sc.nextInt();
        System.out.print("Enter the number of columns: ");
        int cols = sc.nextInt();

        // Creating a 2D array with user-defined size
        int[][] arr = new int[rows][cols];
        int TotalNumber=rows*cols;

        // Taking user input for array elements
        System.out.println("You can Enter "+TotalNumber+" number of the matrix:");
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                arr[i][j] = sc.nextInt();
            }
        }

        // Printing the matrix
        System.out.println("The entered matrix is:");
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                System.out.print(arr[i][j] + " ");
            }
            System.out.println();
        }

        sc.close();
    }
}

File: ARRAY/2D_Array/TwoD_Array.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let rows, cols, matrix = [];

// Function to take user input for matrix dimensions
function askDimensions() {
    rl.question("Enter the number of rows: ", (rowInput) => {
        rows = parseInt(rowInput);
        rl.question("Enter the number of columns: ", (colInput) => {
            cols = parseInt(colInput);
            
            if (isNaN(rows) || isNaN(cols) || rows <= 0 || cols <= 0) {
                console.log("Please enter valid numbers for rows and columns.");
                rl.close();
                return;
            }

            console.log(`You can enter a total of ${rows * cols} elements.`);
            askMatrixInput(0);
        });
    });
}

// Function to take user input for matrix elements row by row
function askMatrixInput(currentRow) {
    if (currentRow === rows) {
        displayMatrix();
        rl.close();
        return;
    }

    rl.question(`Enter ${cols} space-separated numbers for row ${currentRow + 1}: `, (rowData) => {
        let row = rowData.trim().split(" ").map(Number);

        if (row.length !== cols || row.some(isNaN)) {
            console.log(`Please enter exactly ${cols} valid numbers.`);
            askMatrixInput(currentRow);
        } else {
            matrix.push(row);
            askMatrixInput(currentRow + 1);
        }
    });
}

// Function to display the matrix
function displayMatrix() {
    console.log("\nThe entered matrix is:");
    matrix.forEach(row => console.log(row.join(" ")));
}

// Start the input process
askDimensions();
